[Task Manager] adds capacity estimation to the TM health endpoint (#100475)
elastic/kibana: docs/user/production-considerations/task-manager-troubleshooting.asciidoc

[[task-manager-theory-spikes-in-non-recurring-tasks]]
*Theory*:
Spikes in non-recurring and ephemeral tasks are consuming a high percentage of the available capacity

*Diagnosis*:
Task Manager uses ad-hoc non-recurring tasks to load balance operations across multiple {kib} instances.
Additionally, {kib} can use Task Manager to allocate resources for expensive operations by executing an ephemeral task. Ephemeral tasks are identical in operation to non-recurring tasks, but are not persisted and cannot be load balanced across {kib} instances.

Evaluating the preceding health stats, you see the following output under `stats.runtime.value.execution.persistence`:

[source,json]
--------------------------------------------------
{
  "recurring": 88, # <1>
  "non_recurring": 4, # <2>
  "ephemeral": 8 # <3>
},
--------------------------------------------------
<1> 88% of executed tasks are recurring tasks
<2> 4% of executed tasks are non-recurring tasks
<3> 8% of executed tasks are ephemeral tasks

You can infer from these stats that the majority of executions consist of recurring tasks at 88%.

> You can use the `execution.persistence` stats to evaluate the ratio of consumed capacity, but on their own, you should not make assumptions about the sufficency of the availabe capacity.

Reviewer: ```suggestion
You can use the `execution.persistence` stats to evaluate the ratio of consumed capacity, but on their own, you should not make assumptions about the sufficiency of the available capacity.
```

[source,json]
--------------------------------------------------
{
  "recurring": 60, # <1>
  "non_recurring": 30, # <2>
  "ephemeral": 10 # <3>
},
--------------------------------------------------
<1> 60% of executed tasks are recurring tasks
<2> 30% of executed tasks are non-recurring tasks
<3> 10% of executed tasks are ephemeral tasks

You can infer from these stats that even though most executions are recurring tasks, a substantial percentage of executions are non-recurring and ephemeral tasks at 40%.

Evaluate the `load` under `stats.runtime.value`, we see the following:

[source,json]
--------------------------------------------------
{
    "load": { # <2>
        "p50": 70,
        "p90": 100,
        "p95": 100,
        "p99": 100
    }
}
--------------------------------------------------

You can infer from these stats that it is quite common for this {kib} instance to run out of capacity.
Given the high rate of non-recurring and ephemeral tasks, it would be reasonable to assess that there is insufficient capacity in the {kib} cluster to handle the amount of tasks.

> It is important to keep in mind that these stats give you a glimpse at a moment in time, and even though there has been insufficient capacity in recent minutes, this might not be true in other times where fewer non-recurring or ephemeral tasks are used. We recommend tracking these stats over time and identifying the source of these tasks before making sweeping changes to your infrastruture.

Reviewer: ```suggestion
It is important to keep in mind that these stats give you a glimpse at a moment in time, and even though there has been insufficient capacity in recent minutes, this might not be true in other times where fewer non-recurring or ephemeral tasks are used. We recommend tracking these stats over time and identifying the source of these tasks before making sweeping changes to your infrastructure.
```

These rough calculations give you a lower bound to the required throughput, which is _at least_ 410 tasks per minute to ensure recurring tasks are executed, at their scheduled time. This throughput doesn't account for nonrecurring tasks that might have been scheduled, nor does it account for tasks (recurring or otherwise) that might be scheduled in the future.

Given these inferred attributes, it would be safe to assume that a single {kib} instance with default settings **would not** provide the required throughput. It is possible that scaling horizontally by adding a couple more {kib} instances will.

For details on scaling Task Manager, see <<task-manager-scaling-guidance>>.

[[task-manager-health-evaluate-the-capacity-estimation]]
===== Evaluate the Capacity Estimation

Task Manager is constantly evaluating its runtime operations and workload. This enables Task Manager to make rough estimates about the sufficiency of its capacity.

> As the name suggests, these are estimates based on historical data and should not be used as predictions. These estimations should be evaluated along side the detailed <<task-manager-health-monitoring>> stats before making changes to infrastructure. These estimations assumes all {kib} instances are configured identically.

Reviewer: ```suggestion
As the name suggests, these are estimates based on historical data and should not be used as predictions. These estimations should be evaluated alongside the detailed <<task-manager-health-monitoring>> stats before making changes to infrastructure. These estimations assumes all {kib} instances are configured identically.
```

We recommend using these estimations when following the Task Manager <<task-manager-scaling-guidance>>.

> Evaluating the health stats above you can see the following output under `stats.capacity_estimation.value`:

Author: Indeed, it refers to the example output at the top of the page.
We used this approach in the previous sections as well, to avoid making the docs too long an repetitive, but perhaps it's better to copy the output tot he top of each section? 🤔 I'm not sure.

> Evaluating the health stats above you can see the following output under `stats.capacity_estimation.value`:

Reviewer 2: I'd just change the text from "above" to "in the previous example"

[source,json]
--------------------------------------------------
{
  "observed": {
    "observed_kibana_instances": 1, # <1>
    "minutes_to_drain_overdue": 1, # <2>
    "max_throughput_per_minute_per_kibana": 200,
    "max_throughput_per_minute": 200, # <3>
    "avg_recurring_required_throughput_per_minute": 28, # <4>
    "avg_recurring_required_throughput_per_minute_per_kibana": 28,
    "avg_required_throughput_per_minute": 28, # <5>
    "avg_required_throughput_per_minute_per_kibana": 28
  },
  "proposed": {
    "min_required_kibana": 1, # <6>
    "proposed_kibana": 1, # <7>
    "avg_recurring_required_throughput_per_minute_per_kibana": 28,
    "avg_required_throughput_per_minute_per_kibana": 28
  }
}
--------------------------------------------------

> <1> These estimates assume that there is one {kib} instance actively executing tasks

Reviewer 3: To me `observed_kibana_instances` means the number of Kibanas running, while this description says `estimate`. Is this an estimate or an actual number?

Author: It's an estimate based on the observed number of OwnerIds.
There is no way, currently, for Kibana to know how many instances are running, which is frustrating but we have to work with it.
In the code you'll see a bunch of calculations where we use the term "assumed" - as in, we assume this top be the case, but it's not 100% as we're basing this off of the observed number of Kibana. In order for us to produce any kind of estimate for the non-recurring we need some kind of assumption to work off of, and so observing the OwnerIds is the closest we can come to producing an estimate.

Author: I spoke to @chrisronline and he thinks Stack Monitoring could provide us a more reliable number.
I'll file a follow up issue for us to look into that in the near future, so we can improve this incrementally.

> <2> Based on past throughput the overdue tasks in the system could be executed within 1 minute

Reviewer 3: Is there a guideline as to how many `minutes_to_drain_overdue` is too many? Comparing this example to the below example of an overloaded deployment, `1` minute to drain vs `12` minutes to drain doesn't seem too bad?

Author: Not really.
In theory, we shouldn't have any overdue tasks, so anything > 0 is a sign that we are low on capacity.
12 minutes isn't too bad until you realized that during those 12 minutes, additional overdue tasks will build up.

I tried to calrify that in the docs saying:
```
* There appear to be so many overdue tasks that it would take 12 minutes of executions to catch up with that backlog. This does not take into account tasks that might become overdue during those 12 minutes, so while this congestion might be temporary, the system could also remain consistently under provisioned and might never drain the backlog entirely.
```

> <3> Assuming all {kib} instacnes in the cluster are configured the same as this instance, the maximum available throughput is 200 tasks per minute

Reviewer: ```suggestion
<3> Assuming all {kib} instances in the cluster are configured the same as this instance, the maximum available throughput is 200 tasks per minute
```

<4> On average the recurring tasks in the system have historically required a throughput of 28 tasks per minute
<5> On average, regardless of whether they are recurring or otherwise, the tasks in the system have historically required a throughput of 28 tasks per minute
<6> One {kib} instance should be sufficient to run the current recurring workload

> <7> We propose waiting for the workload to change before additional {kib} instances are provisioned

Author: Haha, will do :)
I don't intuitively do this on lists, this is why I need your guidance... 🙏

--------------------------------------------------
<1> These estimates assume that there are two {kib} instance actively executing tasks
<2> The maximum available throughput in the system currently is 400 tasks per minute
<3> Based on past throughput the overdue tasks in the system should be executed within 12 minute
<4> On average the recurring tasks in the system have historically required a throughput of 354 tasks per minute
<5> On average each {kib} instance utilizes 177 tasks per minute of its capacity to execute recurring tasks
<6> On average the tasks in the system have historically required a throughput of 434 tasks per minute
<7> The system estimates that at least two {kib} instances are required to run the current recurring workload
<8> The system recommends provisioning three {kib} instances to handle the workload
<9> Once a third {kib} instance is provisioned, the capacity utilized by each instance to execute recurring tasks should drop from 177 to 118 tasks per minute
<10> Taking into account historical ad-hoc task execution, we estimate the throughput required of each {kib} instance will drop from 217 task per minute to 145, once a third {kib} instance is provisioned

> 

Reviewer 2: Add ending period to the items in this list?

Evaluating by these estimates, we can infer some interesting attributes of our system:

* These estimates are produced based on the assumption that there are two {kib} instances in the cluster. This number is based on the number of {kib} instances actively executing tasks in recent minutes. At times this number might fluctuate if {kib} instances remain idle, so validating these estimates against what you know about the system is recommended.

> * There appear to be so many overdue tasks that it would take 12 minutes of executions to catch up with that backlog. This does not take into account tasks that might become overdue during those 12 minutes, so while this congestion might be temporary, the system could also remain consistently under provisioned and might never drain the backlog entirely.

Reviewer 2: ```suggestion
* There appear to be so many overdue tasks that it would take 12 minutes of executions to catch up with that backlog. This does not take into account tasks that might become overdue during those 12 minutes. Although this congestion might be temporary, the system could also remain consistently under provisioned and might never drain the backlog entirely.
```

> * Evaluating the recurring tasks in the workload the system requires a throughput of 354 tasks per minute on average to execute tasks on time, which is lower then the estimated maximum throughput of 400 tasks per minute. Once we take into account historical throughput though, we estimate that the required throughput at 434 tasks per minute. This suggests that, historically, approximately 20% of tasks have been ad-hoc non-recurring tasks, the scale of which are harder to predict than recurring tasks.

Reviewer 2: ```suggestion
* Evaluating the recurring tasks in the workload, the system requires a throughput of 354 tasks per minute on average to execute tasks on time, which is lower then the estimated maximum throughput of 400 tasks per minute. Once we take into account historical throughput though, we estimate the required throughput at 434 tasks per minute. This suggests that, historically, approximately 20% of tasks have been ad-hoc non-recurring tasks, the scale of which are harder to predict than recurring tasks.
```

> * Evauating the recurring tasks in the workload the system requires a throughput of 354 tasks per minute on average to execute tasks on time, which is lower then the estimated maximum throughput of 400 tasks per minute. Once we take into account historical throughpout though, we estimate that the required throughput at 434 tasks per minute. This suggests that, historically, approximately 20% of tasks have been ad-hoc non-recurring tasks, the scale of which are harder to predict than recurring tasks.

Reviewer: ```suggestion
* Evaluating the recurring tasks in the workload the system requires a throughput of 354 tasks per minute on average to execute tasks on time, which is lower then the estimated maximum throughput of 400 tasks per minute. Once we take into account historical throughput though, we estimate that the required throughput at 434 tasks per minute. This suggests that, historically, approximately 20% of tasks have been ad-hoc non-recurring tasks, the scale of which are harder to predict than recurring tasks.
```

> You can infer from these estimates that the capacity in the current system is insufficient and at least one additional {kib} instance is required in order to keep up with the workload.

Reviewer 2: ```suggestion
You can infer from these estimates that the capacity in the current system is insufficient and at least one additional {kib} instance is required to keep up with the workload.
```
